scripts/mbedtls_framework/code_wrapper/psa_wrapper.py

```diff
@@ -0,0 +1,257 @@
+#!/usr/bin/env python3
+"""Generate wrapper functions for PSA function calls.
+"""
+
+# Copyright The Mbed TLS Contributors
+# SPDX-License-Identifier: Apache-2.0 OR GPL-2.0-or-later
+
+import argparse
+import itertools
+import os
+from typing import Any, Iterator, List, Dict, Collection, Optional, Tuple
+
+from mbedtls_framework import build_tree
+from mbedtls_framework import c_parsing_helper
+from mbedtls_framework import c_wrapper_generator
+from mbedtls_framework import typing_util
+
+from mbedtls_framework.code_wrapper.psa_buffer import BufferParameter
+from textwrap import dedent
+
+DEFAULTS = {
```

Comment:
I have changed that but I would like to note my objection.

This is an immutable key:value store acting as a controlled user input. It does not need to be modified in any way. Dictionary is the best and more efficient way to handle it. 

What benefit are we getting by wrapping the said dictionary onto a Data-Class, other adding complexity?

### Dict design
~~~~~~

config = {key1: value1
          key2: value2
          ....}
          
def some_method(config: Dict[str, Any]= config) -> None:
~~~~~~ 

Here we have a global key:value string store, where the method takes the pointer as an argument.
The data is not modified in any way, and used when needed using that pointer.

### Proposed design
~~~~~~
class CFG.__dict__ =  {key1: value1
                       key2: value2
                       ....}
def some_method(config: CFG = CFG()) -> None:
~~~~~~

each time the some_method() is called, the following things happen.

- class CFG is memcopied onto a new location.
- The pointer of the copy is passed as the default argument of some_method.
- The pointer is added to the Garbage collector pool.
- Later on the garbage collector has to poll if that is still active and cleanup.

I fail to see why the second is prefered for configuration data that does not need to be manipulated, and do not require any special handling to be formatted. The only benefit I see is that it passes the contrains we set to ourselves by the limits of mypi.

Reply:
> This is an immutable key:value store acting as a controlled user input.

Not just that, **the set of keys is always the same**. That goes far from the basic properties of a dict, which have a variable and mutable set of keys. Python has a data structure exactly for that: a [namedtuple](https://docs.python.org/3/library/collections.html#collections.namedtuple). A namedtuple works well as long as you only want to bundle data together, without defining any special behavior on the class. If you want additional methods, a custom class tends to be simpler than extending the namedtuple.

> each time the some_method() is called, the following things happen.
>
> * class CFG is memcopied onto a new location.

No. The argument is passed by reference. Python doesn't have a concept of passing by value. Python implementations can pass objects by value _if_ the object is immutable (otherwise that would violate the semantics) and they decide that it's an optimization; I believe CPython does that for small integers. But in any case, a namedtuple or a class that's defined manually with similar properties would not be copied in any way that's different from a dict here.

> I fail to see why the second is prefered for configuration data that does not need to be manipulated

Not that this matters anyway, since we don't have particular performance or memory usage concerns in this script. The primary concern is readability and maintainability (the two are closely related). And in this respect, a namedtuple or similar is a clear win:

* A known set of elements, so it's easy for readers to understand what field names are valid, and static checkers can detect mistakes with field names.
* Each field's value has its own type of element. When fields are fields, it's easy to understand: `foo.bar` is an integer, `foo.qux` is a list of strings, etc. With dictionaries, you'd see `foo[field_name]` and you'd have no idea what that is.
* Immutable data structure, so no risk of modifying when unintended. (But even if the structure was mutable, a [`SimpleNamespace`](https://docs.python.org/3/library/types.html#types.SimpleNamespace) or a custom class would be preferable to a dictionary when the set of fields is always the same.)

> The only benefit I see is that it passes the contrains we set to ourselves by the limits of mypi.

You're switching cause and effect here. The data structure follows some rules, which are completely hidden if you stuff it in a `dict`. When you don't hide the rules, mypi helps you. When you hide the rules, you run into trouble with mypi: mypi is rightfully telling you that you're doing something wrong.